streams: forward errors correctly for duplexPair endpoints

aelhor opened the pull request:
fix the duplexPair implementation so that when one side is destroyed with an error, the other side also receives the error or a close event as appropriate.

previous behavior caused sideA to never emit an 'error' or 'close' when sideB errored, which prevented users from observing or handling the paired stream failure.

Fixes: https://github.com/nodejs/node/issues/61015

<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

lpinca commented on test/parallel/test-duplex-error.js:
```diff
@@ -0,0 +1,30 @@
+
+let sideAErrorReceived = false;
+let sideBErrorReceived = false;
+
```
```suggestion
```

lpinca commented on test/parallel/test-duplex-error.js:
```diff
@@ -0,0 +1,30 @@
+
+// Use common.mustCall inside the listeners to ensure they trigger
+sideA.on('error', common.mustCall((err) => {
+  sideAErrorReceived = true;
```
```suggestion
```

You might want to check that the error is the expected one instead.

lpinca commented on test/parallel/test-duplex-error.js:
```diff
@@ -0,0 +1,30 @@
+}));
+
+sideB.on('error', common.mustCall((err) => {
+  sideBErrorReceived = true;
```
```suggestion
```

Ditto.

aelhor commented:
Hi @lpinca @mcollina @Renegade334 👋  
Just a gentle ping in case this PR slipped through the cracks.  
No rush at all — I’d really appreciate any feedback or guidance when you have time.  
Thanks!

aelhor commented:
@mcollina 
I have updated the commit messages to use the correct stream: prefix and squashed the review suggestions and lint fixes into the main commits to keep the history clean. Ready for another look!

aelhor commented:
CI is now green after the latest fix. Since this addresses the previous failure, would it make sense to re-evaluate the semver-major label to see if it’s still applicable?

StefanStojanovic commented:
I've retrigged CI and all passed, reran the failed GHA as well.

dylanreniers commented:
Hi all,

A recent upgrade to 24.16.0 has caused me some issues with my CI pipeline (Jenkins, running on Amazon Linux 2023). During the pipeline it tries to install playwright for running angular tests (using vitest browser). It downloads chromium successfully and then it hangs trying to extract the package. This does not happen in 24.14.0

I have asked my coding agent to help me find the cause of this issue and it pointed me here with the following explanation:

```
Playwright's extract-zip (via yauzl) creates a stream pipeline: read zip entry → pipe to write stream → next entry. The stream pipeline behavior change in 24.15.0 is the most likely root cause — if it
  changed how a pipeline signals completion or handles certain stream states, it could cause the extraction pipeline to stall waiting for a signal that never comes. The duplexPair destruction change in
  24.16.0 may compound it.

  This also means 24.15.0 likely has the same hang — your issue is probably not specific to 24.16.0 but was introduced in 24.15.0. You might want to test 24.15.0 to confirm before assuming 24.16.0 is the
  In 24.15.0:
  - "Stream pipeline — Preserves error over AbortError in pipeline" — changed how pipeline completion/errors are handled
  - "Compression improvements — Better spec compliance for Web Compression streams" — changed how compression streams behave
```

I am not familiar enough with the internals of nodejs to help out with this, but I did want to raise it as perhaps other people would also run into this issue. For now we can stay on 24.14.0, so no big deal, but something to keep in mind.

Renegade334 commented:
> Hi all,
> 
> A recent upgrade to 24.16.0 has caused me some issues with my CI pipeline (Jenkins, running on Amazon Linux 2023). During the pipeline it tries to install playwright for running angular tests (using vitest browser). It downloads chromium successfully and then it hangs trying to extract the package. This does not happen in 24.14.0
> 
> I have asked my coding agent to help me find the cause of this issue and it pointed me here with the following explanation:
> 
> ```
> Playwright's extract-zip (via yauzl) creates a stream pipeline: read zip entry → pipe to write stream → next entry. The stream pipeline behavior change in 24.15.0 is the most likely root cause — if it
>   changed how a pipeline signals completion or handles certain stream states, it could cause the extraction pipeline to stall waiting for a signal that never comes. The duplexPair destruction change in
>   24.16.0 may compound it.
> 
>   This also means 24.15.0 likely has the same hang — your issue is probably not specific to 24.16.0 but was introduced in 24.15.0. You might want to test 24.15.0 to confirm before assuming 24.16.0 is the
>   In 24.15.0:
>   - "Stream pipeline — Preserves error over AbortError in pipeline" — changed how pipeline completion/errors are handled
>   - "Compression improvements — Better spec compliance for Web Compression streams" — changed how compression streams behave
> ```
> 
> I am not familiar enough with the internals of nodejs to help out with this, but I did want to raise it as perhaps other people would also run into this issue. For now we can stay on 24.14.0, so no big deal, but something to keep in mind.

This PR is unrelated to that issue.
